File: .github/actions/capture-server-diagnostics/sanitize-logs.py
#!/usr/bin/env python3
"""Sanitize server logs by redacting PII before artifact upload."""
import sys
import json
import re

def sanitize_logs(input_file, output_file):
    """Read raw logs, redact PII, and write sanitized logs."""
    try:
        with open(input_file) as f:
            try:
                logs = json.load(f)
            except json.JSONDecodeError:
                # If JSON is malformed, write empty array
                logs = []

        # Sanitize each log entry: redact emails, usernames, tokens, and sensitive fields
        for entry in logs if isinstance(logs, list) else []:
            if not isinstance(entry, dict):
                continue
            msg = entry.get('message', '')
            if not isinstance(msg, str):
                continue

            # Redact email-like patterns
            msg = re.sub(r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}', '[redacted-email]', msg)
            # Redact common credential patterns (tokens, passwords in logs)
            msg = re.sub(r'(?:token|password|auth|Bearer|bearer)\s*[:=]\s*["\']?[^\s"\']+["\']?', '[redacted-credential]', msg, flags=re.IGNORECASE)
            # Redact usernames in quotes
            msg = re.sub(r'"(?:username|user_id|user|actor)":\s*"[^"]*"', '"[redacted-field]": "[redacted-value]"', msg)
            # Redact numeric IDs that might be sensitive
            msg = re.sub(r'"id":\s*"[a-z0-9]{25,}"', '"id": "[redacted-id]"', msg)

            entry['message'] = msg

        # Write sanitized logs
        with open(output_file, 'w') as f:
            json.dump(logs, f, indent=2)
        return True
    except Exception:
        # On error, write empty array rather than fail the diagnostics step
        with open(output_file, 'w') as f:
            json.dump([], f)
        return True

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print("usage: sanitize-logs.py <input-file> <output-file>", file=sys.stderr)
        sys.exit(1)
    input_file, output_file = sys.argv[1], sys.argv[2]
    if not sanitize_logs(input_file, output_file):
        sys.exit(1)

File: .github/actions/capture-server-diagnostics/sanitize-logs.js
#!/usr/bin/env node
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

const fs = require('node:fs');
const path = require('node:path');

const EMAIL = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;
const CREDENTIAL = /\b(?:token|password|auth)\s*[:=]\s*["']?[^\s"']+["']?|\bbearer\s+["']?[^\s"']+["']?/gi;
const SENSITIVE_FIELD = /"(?:username|user_id|user|actor)"\s*:\s*"[^"]*"/gi;
const LONG_ID = /"id"\s*:\s*"[a-z0-9]{25,}"/gi;

function sanitizeMessage(message) {
    return message.
        replace(EMAIL, '[redacted-email]').
        replace(CREDENTIAL, '[redacted-credential]').
        replace(SENSITIVE_FIELD, '"[redacted-field]": "[redacted-value]"').
        replace(LONG_ID, '"id": "[redacted-id]"');
}

function sanitizeLogs(logs) {
    if (!Array.isArray(logs)) {
        return [];
    }

    return logs.map((entry) => {
        if (!entry || typeof entry !== 'object' || Array.isArray(entry) || typeof entry.message !== 'string') {
            return entry;
        }
        return {...entry, message: sanitizeMessage(entry.message)};
    });
}

function writeJson(outputFile, value) {
    fs.mkdirSync(path.dirname(outputFile), {recursive: true});
    fs.writeFileSync(outputFile, `${JSON.stringify(value, null, 2)}\n`, 'utf8');
}

function sanitizeFile(inputFile, outputFile) {
    try {
        const logs = JSON.parse(fs.readFileSync(inputFile, 'utf8'));
        writeJson(outputFile, sanitizeLogs(logs));
    } catch {
        // Diagnostics are best-effort. Invalid or unreadable input must never
        // upload raw logs or fail the E2E job.
        writeJson(outputFile, []);
    }
}

function main(argv = process.argv.slice(2)) {
    if (argv.length !== 2) {
        console.error('usage: sanitize-logs.js <input-file> <output-file>');
        return 1;
    }
    sanitizeFile(argv[0], argv[1]);
    return 0;
}

module.exports = {main, sanitizeFile, sanitizeLogs, sanitizeMessage};

if (require.main === module) {
    process.exitCode = main();
}
